PHP/PHP Source.sublime-syntax

```diff
@@ -1217,110 +1267,152 @@ contexts:
+            3: punctuation.definition.tag.begin.php
+            4: entity.name.tag.heredoc.php
+            5: punctuation.definition.tag.end.php
           embed: scope:source.sql
           embed_scope: meta.embedded.sql source.sql
           escape: (?=^\s*SQL\b)
-        - match: <<<\s*(JAVASCRIPT)\s*$\n?
-          scope: punctuation.section.embedded.begin.php punctuation.definition.string.php
+        - match: (<<<)\s*(()(JAVASCRIPT)())\s*$\n?
           captures:
             1: keyword.operator.heredoc.php
+            2: meta.string.heredoc.php meta.tag.heredoc.php
+            3: punctuation.definition.tag.begin.php
+            4: entity.name.tag.heredoc.php
+            5: punctuation.definition.tag.end.php
           embed: heredoc-javascript
-          embed_scope: meta.embedded.js source.js
+          embed_scope: meta.embedded.js
           escape: (?=^\s*JAVASCRIPT\b)
-        - match: <<<\s*('JAVASCRIPT')\s*$\n?
-          scope: meta.embedded.js punctuation.section.embedded.begin.php punctuation.definition.string.php
+        - match: (<<<)\s*((')(JAVASCRIPT)('))\s*$\n?
           captures:
-            1: keyword.operator.nowdoc.php
+            1: keyword.operator.heredoc.php
+            2: meta.string.heredoc.php meta.tag.heredoc.php
+            3: punctuation.definition.tag.begin.php
+            4: entity.name.tag.heredoc.php
+            5: punctuation.definition.tag.end.php
           embed: scope:source.js
           embed_scope: meta.embedded.js source.js
           escape: (?=^\s*JAVASCRIPT\b)
-        - match: (<<<)\s*(JSON)\s*$\n?
-          scope: punctuation.section.embedded.begin.php punctuation.definition.string.php
+        - match: (<<<)\s*(()(JSON)())\s*$\n?
           captures:
             1: keyword.operator.heredoc.php
+            2: meta.string.heredoc.php meta.tag.heredoc.php
+            3: punctuation.definition.tag.begin.php
+            4: entity.name.tag.heredoc.php
+            5: punctuation.definition.tag.end.php
           embed: heredoc-json
-          embed_scope: meta.embedded.json source.json
+          embed_scope: meta.embedded.json
           escape: (?=^\s*JSON\b)
-        - match: (<<<)\s*('JSON')\s*$\n?
-          scope: meta.embedded.json punctuation.section.embedded.begin.php punctuation.definition.string.php
+        - match: (<<<)\s*((')(JSON)('))\s*$\n?
           captures:
-            1: keyword.operator.nowdoc.php
+            1: keyword.operator.heredoc.php
+            2: meta.string.heredoc.php meta.tag.heredoc.php
+            3: punctuation.definition.tag.begin.php
+            4: entity.name.tag.heredoc.php
+            5: punctuation.definition.tag.end.php
           embed: scope:source.json
           embed_scope: meta.embedded.json source.json
           escape: (?=^\s*JSON\b)
-        - match: <<<\s*(CSS)\s*$\n?
-          scope: punctuation.section.embedded.begin.php punctuation.definition.string.php
+        - match: (<<<)\s*(()(CSS)())\s*$\n?
```

Comment:
Interesting technique to have empty capture groups to be able to copy/paste the scopes. Is it worth using a lookahead to check for both quotes / no quotes and make the capture groups for the quotes optional to avoid needing the almost-duplicate match pattern?

Reply:
Good catch. Maybe we could use the following to reduce duplication?

```suggestion
        - match: (<<<)\s*(('?)(CSS)(\3?))\s*$\n?
```

I made the closing punctuation optional, too, to help with not breaking syntax in the middle of typing. But it could also be obligatory.

Reply:
Doesn't using a capture group reference in a non-pop pattern cause Oniguruma to be used? Can you double check the Compatibility Tests please?

Reply:
It does indeed trigger it and mark the rule incompatible. But it does so, too, if backrefs are used in popping patterns included in a pushed context. ST just doesn't complain then.

As PHP uses single quoted tags only, we could also use:

```suggestion
        - match: (<<<)\s*(('?)(CSS)('?))\s*$\n?
```

... to lazily handle quoted CSS.

Reply:
We shouldn't merge those patterns atm, because one version supports PHP variable interpolation and the other does not. (see lines 1394 vs. 1406)

We may provide intermediate inherited syntaxes in future for those supporting interpolation. This is actually what I already prepared in https://github.com/deathaxe/sublime-Packages/tree/wip/php (needs some update once this one is merged, though).

Reply:
somehow I missed that, thanks for spotting it :)

Reply:
The `escape: (?=^\s*JSON\b)` can use backreference too actually.